LSPS4/README.md

```diff
@@ -0,0 +1,1889 @@
+normally due to a race condition that changes the liquidity
+towards the client:
+
+* MUST do one of the following:
+  * (preferred) add the PPP back to the PPP Map under a
+    new PPP identifier and resumes its hold timer.
+    It re-sends `lsps4.you_have_new_ppp` under the new PPP
+    identifier.
+    If its hold timer has already expired at this point, the
+    LSP MUST NOT do this, but MUST do the alternative below
+    instead.
+  * (unpreferred) NOT add the PPP back to the PPP Map and
+    fail the payment with `temporary_channel_failure`.
+
+> **Rationale** Most node implementations, once you have
+> indicated that the HTLC is to be forwarded "normally", will
+> take over processing, and if it then turns out that there
+> is insufficient liquidity, will automatically fail the
+> HTLC with no way to re-handle the HTLC.
+> Thus, the "unpreferred" branch is allowed (it is the only
+> behavior such node implementations can do), but is still
+> marked as unpreferred as it increases payment failure (the
+> payer might give up retrying).
+> The hope is that node implementations will eventually be
+> modified so that the preferred branch can be taken instead.
+
+Regardless, normal BOLT-specified handling on the client
+would still work despite this issue:
+
+* If the payment is a single part:
+  * If the preferred branch is taken by the LSP, the
+    client gets notified and can retry the
+    `lsps4.forward_ppps` call, hopefully now with the updated
+    information on available liquidity and onchain fees
+    allowing the call to correctly report
+    `insufficient_liquidity`.
+  * If the unpreferred branch is taken by the LSP, the
+    payer may still retry, and the client would again be
+    notified of the new PPP, with the same result as above.
+* If the payment is multipart:
+  * Any parts that were successfully forwarded will eventually
+    time out with the remaining parts undelivered, and a
+    BOLT-compliant client will `mpp_timeout` the delivered
+    parts.
+  * Any parts that were not successfully forwarded will be
+    re-added to the client copy of the PPP Map (if the LSP
+    takes the preferred branch) for re-evaluation by the
+    client, or might eventually be retried by the payer
+    (if the LSP takes the unpreferred branch).
+
+#### Failing Pending Payment Parts
+
+A pending payment part can be failed by the client using the
+`lsps4.fail_ppp` API.
+
+For example, suppose that the payer attempts a multipath
+payment to the client via the LSP.
+Some of the parts arrive at the LSP and are converted to PPPs.
+However, other parts fail to reach the LSP, and eventually
+the payer gives up.
+
+The client, as a "good citizen", SHOULD fail the multipath
+parts that reached the LSP using an `mpp_timeout` error, in
+order to free up network resources.
+
+`lsps4.fail_ppp` accepts the parameters:
+
+```JSON
+{
+  "ppp_id": "id#abcdefghijklmnop",
+  "details": {
+    "type": "htlc",
```

Comment:
Type is redundant here because the LSP already knows the type through the `ppp_id`.

Reply:
Better to be explicit than implicit.

Reply:
Don't see any reason for that. Looks to me like a "duplicated code" code smell.

Reply:
Done.